### ISSUE TITLE
https://github.com/gajus/eslint-plugin-jsdoc (recommended)

### ISSUE THREAD

**Issue by camc314** [web-infra-dev/oxc#1170]

**Comment by Boshen** [comment]
The jsdoc parser is incomplete so we'll have to iteratively expand the jsdoc functionality as we add more rules.

**Comment by leaysgur** [comment]
Updates... 📝 

- JSDoc finding APIs have implemented in #2437 
- JSDoc parsing is still under way

`eslint-plugin-jsdoc` rules can be divided into 4 categories.

- check every all comments(1/53 rule)
- check every JSDoc comments(27/53 rules)
- check every JSDoc comments + additional check with attaching node(3/53 rules)
- check specific nodes with attached JSDoc comments(22/53 rules)

(IMO, difficulty is from easier:top to harder:bottom)

**Comment by leaysgur** [comment]
I have been working on this plugin for the past 6 months.
As a personal goal, I aimed to implement rules that are recommended, not stylistic, and not type-checked.

At this time,

- Mostly achieved that
  - Except for `check-param-names`, which is really complex and out of control 🫠
- I am feeling somewhat burnt out

Therefore, I am considering stepping away from this issue.

If anyone is interested in taking on this challenge, feel free to ask me questions. 😉

**Comment by Boshen** [comment]
@leaysgur Thank you so much for your contribution ❤️ 

My intention was to implement a jsdoc parser with you, but you have blown me away with your endurance.

Please take some rest. If you feel like working on some other parts of the project, feel free to DM on twitter or discord.

**Comment by thernstig** [comment]
Will the new [Oxlint JS Plugins](https://oxc.rs/blog/2025-10-09-oxlint-js-plugins.html) work fine with the native ESLint plugin? I am guessing so, but the main reason I am raising this here is that I believe it is impossible for Oxc to natively implement all rules considering the speed at of which the JSDoc plugin/config moves. (The entire reason you solved the JS plugins support).

I am wondering what will happen with these half-implemented Rust-native plugins such as jsdoc now?

**Comment by overlookmotel** [comment]
We're discussing how to deal with `jsdoc` plugin re: JS plugins on #15413.

**Comment by minsoo-web** [comment]
## Proposal: Implement `jsdoc/valid-types`

Hi! I'd like to implement the `jsdoc/valid-types` rule. Before starting, I wanted to share the approach and get feedback.

### Why this rule is significant

Unlike other jsdoc rules, `valid-types` requires a **JSDoc type expression parser** — `oxc_jsdoc` currently extracts types as raw strings (`JSDocTagTypePart::parsed()`) with no grammar validation. This means the implementation has two parts:

### Proposed approach: 2 PRs

**PR1: JSDoc type expression parser in `oxc_jsdoc`**
- Validation-only recursive descent parser (no AST, just `Result<(), Error>`)
- Supports all 4 modes: `jsdoc`, `typescript`, `closure`, `permissive`
- Namepath validator (`module:foo.Bar#baz` syntax)
- New public API: `validate_jsdoc_type(source, mode, span_offset)`
- Add `pub fn body()` accessor to `JSDocTag` (for `@borrows` parsing)
- ~1500-2500 lines

**PR2: `jsdoc/valid-types` linter rule in `oxc_linter`**
- `run_once` pattern (iterate all JSDoc comments)
- Validates: type expressions, namepaths, parameter names, structural rules (`@borrows as`, inline tags)
- Adds `settings.jsdoc.mode` and `settings.jsdoc.structuredTags` to `JSDocPluginSettings` (currently commented out as "Not planning to support for now")
- `allowEmptyNamepaths` option
- ESLint test case porting (~135 cases)
- ~2000-3500 lines

### Key design decisions

1. **Validation-only parser** — No AST needed since the rule only checks pass/fail. Zero heap allocation, smaller API surface. AST can be added later if `check-types` or `no-undefined-types` rules need it.
2. **Module placement** — `crates/oxc_jsdoc/src/type_parser/` as a sibling to `src/parser/` (independent from comment parser)
3. **Error type** — `JSDocTypeErrorKind` enum (not `String` or `OxcDiagnostic`), keeping `oxc_jsdoc` free of `oxc_diagnostics` dependency
4. **Mode-expression matrix** — Verified against `jsdoc-type-pratt-parser` source. Many expressions (`?T`, `!T`, `T=`, `T[]`, `Type.<T>`) are in `baseGrammar` and valid in all modes.

### Questions for maintainers

1. Is the 2-PR split (infra → rule) acceptable, or would you prefer a single PR?
2. Any concerns about adding `mode` and `structuredTags` to `JSDocPluginSettings`? (These are currently deferred with TODO comments)
3. Is `src/type_parser/` the right location in `oxc_jsdoc`, or would you prefer it under `src/parser/type_parser/`?
4. For the validation-only approach — is it OK to defer AST construction to a future PR if needed?

Looking forward to feedback before I start implementation!

**Comment by camc314** [comment]
@minsoo-web can you make a new issue to discuss this rule?

**Comment by sakgoyal** [comment]
That looked like an AI comment. Idk if that's a human

**Comment by minsoo-web** [comment]
> [@minsoo-web](https://github.com/minsoo-web) can you make a new issue to discuss this rule?

I created!

https://github.com/oxc-project/oxc/issues/20744

**Comment by minsoo-web** [comment]
> That looked like an AI comment. Idk if that's a human

yes, that comment was generated with claude code. I found this issue while migrating from eslint to oxlint and wanted to propose an approach before starting. 

but I should have reviewed it more carefully before posting. Sorry about that.
